### DISCUSSION TITLE
How can I view a defined value at build time

### DISCUSSION THREAD

**Discussion by alextrical** [stm32duino/Arduino_Core_STM32#1864] [Q&A] [answered]
Hi all

I'm currently having issues with the Virtual EEPROM not working for a STM32G0B1CBT with 128Kb flash
As I need to get this working, it would be nice to be able to see the values of: 
FLASH_END
FLASH_PAGE_SIZE

How do we go about debugging to print these values out? I have tried using #error and Serial print, to no avail.
Any advice would be greatly appreciated, as I've made no progress on this issue in months.

Many thanks

**Comment by fpistm** [comment]
#warning in the eeprom library code.

**Reply by alextrical** [reply]
Please can you explain how to call a variable it can output.
I have tried to do this, however I cannot see a way to get '#warning' to return a value for a variable such as FLASH_END
FLASH_PAGE_SIZE, It looks to only return a string.

For example placing "#warning FLASH_END" at line 91 of "stm32_eeprom.c" results in the following being reported in the Arduino IDE `/home/alextrical/.arduino15/packages/STMicroelectronics/hardware/stm32/2.3.0/libraries/EEPROM/src/utility/stm32_eeprom.c:92:2: warning: #warning FLASH_BASE_ADDRESS [-Wcpp]
   92 | #warning FLASH_BASE_ADDRESS
      |  ^~~~~~~`

**Comment by fpistm** [comment, marked as answer]
Well in stm32_eeprom.c add:
`#include "utils.h"`

Then after `FLASH_END `definition:
`#pragma message("The value of FLASH_END: " XSTR(FLASH_END))`


Anyway you will not have exact value as it can depends on ll function, example:
```
EEPROM\src\utility\stm32_eeprom.c:75:9: note: '#pragma message: The value of FLASH_END: ((uint32_t)(0x08000000UL + (((((uint32_t)(((LL_GetFlashSize() * 1024) / 0x00000800U) - 1)) +1) * 0x00000800U))-1))'
   75 | #pragma message("The value of FLASH_END: " XSTR(FLASH_END))
      |         ^~~~~~~
```
If you want exact vaule then simply add a method to get it in the c file and export it then display it with Serial at sketch level.

**Reply by alextrical** [reply]
Thank you, I can confirm that `#pragma message` is exactly what i was looking for. Using that I have been able to identify what the values are being set to, even if it doesn't fully resolve to a number as you stated. It still helps to track down whats causing the virtual eeprom to not work on the STM32G0B1CBT MCU line.

So far I have retrieved the following information
```
#pragma message: FLASH_BANK_NUMBER: (0x1UL << (15U))
#pragma message: FLASH_BANK_1: (0x1UL << (2U))
#pragma message: FLASH_BANK_2: (0x1UL << (15U))
#pragma message: FLASH_SECTOR_TOTAL: FLASH_SECTOR_TOTAL
#pragma message: FLASH_DATA_SECTOR: FLASH_DATA_SECTOR
#pragma message: FLASH_BASE_ADDRESS: ((uint32_t)((((uint32_t)((0x08000000UL) + (((((uint32_t)(((LL_GetFlashSize() * 1024) / 0x00000800U) - 1)) +1) * 0x00000800U))-1)) + 1) - 0x00000800U))
#pragma message: FLASH_PAGE_NUMBER: ((uint32_t)(((LL_GetFlashSize() * 1024) / 0x00000800U) - 1))
#pragma message: FLASH_PAGE_SIZE: 0x00000800U
#pragma message: FLASH_END: ((uint32_t)((0x08000000UL) + (((((uint32_t)(((LL_GetFlashSize() * 1024) / 0x00000800U) - 1)) +1) * 0x00000800U))-1))
#pragma message: EEPROM_RETRAM_MODE: EEPROM_RETRAM_MODE
#pragma message: DATA_EEPROM_END: DATA_EEPROM_END
#pragma message: FLASH_BASE: (0x08000000UL)
#pragma message: DATA_EEPROM_BASE: DATA_EEPROM_BASE
```

I will continue to try and track down the source of the bug, but this at least seems to indicate the definition thinks there are 2 memory banks, even though this MCU only has 1. Now to work out how to override that, but being able to debug the code is a great step forward ^_^

Have a great evening.

**Reply by bxparks** [reply]
Wow, I've been using C++ on and off for 30 years, and I never knew about that `#pragma message()`. Would have saved me a lot of time in a few cases.

**Reply by alextrical** [reply]
> If you want exact vaule then simply add a method to get it in the c file and export it

Sorry to be a pain, but could i please ask for a example or a pointer to the relevant reading material.
I have tried using `extern` to expose the value. 
along with constructing the following
`uint8_t read_FLASH_BASE_ADDRESS()
{
  return FLASH_BASE_ADDRESS;
}`
